google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/changestream/ChangeStreamStateMachine.java

```diff
@@ -0,0 +1,580 @@
+              "AWAITING_CELL_VALUE: Can't handle a Heartbeat in the middle of building a SetCell.");
+        }
+
+        @Override
+        State handleCloseStream(ReadChangeStreamResponse.CloseStream closeStream) {
+          throw new IllegalStateException(
+              "AWAITING_CELL_VALUE: Can't handle a CloseStream in the middle of building a SetCell.");
+        }
+
+        @Override
+        State handleMod(ReadChangeStreamResponse.DataChange dataChange, int index) {
+          validate(
+              0 <= index && index <= dataChange.getChunksCount() - 1,
+              "AWAITING_CELL_VALUE: Index out of bound.");
+          ReadChangeStreamResponse.MutationChunk chunk = dataChange.getChunks(index);
+          validate(
+              chunk.getMutation().hasSetCell(),
+              "AWAITING_CELL_VALUE: Current mod is not a SetCell.");
+          Mutation.SetCell setCell = chunk.getMutation().getSetCell();
+          numCellChunks++;
+          builder.cellValue(setCell.getValue());
+          // Case 1: Current SetCell is chunked.
+          if (chunk.hasChunkInfo()) {
+            validate(
+                chunk.getChunkInfo().getChunkedValueSize() > 0,
+                "AWAITING_CELL_VALUE: Chunked value size must be positive.");
+            actualTotalSizeOfChunkedSetCell += setCell.getValue().size();
+            // If it's the last chunk of the chunked SetCell, finish the cell.
+            if (chunk.getChunkInfo().getLastChunk()) {
+              builder.finishCell();
+              validate(
+                  actualTotalSizeOfChunkedSetCell == chunk.getChunkInfo().getChunkedValueSize(),
+                  "Chunked value size in ChunkInfo doesn't match the actual total size. "
+                      + "ChunkInfo: "
+                      + chunk.getChunkInfo().getChunkedValueSize()
+                      + "; actual total size: "
+                      + actualTotalSizeOfChunkedSetCell);
+              return checkAndFinishMutationIfNeeded(dataChange, index + 1);
+            }
+            // If this is not the last chunk of a chunked SetCell, then this must be the last mod of
+            // the current response, and we're expecting the rest of the chunked cells in the
+            // following ReadChangeStream response.
+            validate(
+                index == dataChange.getChunksCount() - 1,
+                "AWAITING_CELL_VALUE: Current SetCell is not chunked.");
+            return AWAITING_CELL_VALUE;
+          }
+          // Case 2: Current SetCell is not chunked.
+          builder.finishCell();
+          return checkAndFinishMutationIfNeeded(dataChange, index + 1);
+        }
+      };
+
+  /**
+   * A state that represents a completed change stream record. It prevents new change stream records
+   * from being read until the current one has been consumed.
+   *
+   * <dl>
+   *   <dt>Valid exit states:
+   *   <dd>{@link ChangeStreamStateMachine#AWAITING_NEW_STREAM_RECORD}.
```

Comment:
how will this exit to AWAITING_NEW_STREAM_RECORD?

Reply:
The user is supposed to call `consumeChangeStreamRecord()` after the `hasCompleteChangeStreamRecord()` succeeds. `consumeChangeStreamRecord` will then call `reset()` and reset the currentState to `AWAITING_NEW_STREAM_RECORD`.

So this state itself won't transit to `AWAITING_NEW_STREAM_RECORD`. I updated the comments.